import pandas as pd
import os.path as path

# ==============================================================================
# Configuration
# ==============================================================================
CONFIG = pd.read_csv('config.tsv', index_col=False, sep='\t')

REPORT_DIR = 'Reports'
FASTQ_DIR = 'FASTQs'
TRIM_DIR = 'Trimmed'
ALIGN_DIR = 'Aligned'
PEAK_DIR = 'Peaks'

SAMPLES = CONFIG["Sample"].tolist()
BWT2_IDX = "/mnt/work1/data/iGenomes/human/hg38/Sequences/WholeGenomeFasta/genome.fa"
READS = [1, 2]

# ==============================================================================
# Meta Rules
# ==============================================================================
rule all:
    input:
        # FastQC reports
        expand(
            path.join(FASTQ_DIR, "{sample}_R{read}_fastqc.{ext}"),
            sample=SAMPLES, read=READS, ext=["html", "zip"]
        ),
        # Trimming reports
        expand(
            path.join(REPORT_DIR, "{sample}_R{read}.trimming_report.txt"),
            sample=SAMPLES, read=READS
        ),
        expand(
            path.join(TRIM_DIR, "{sample}_R{read}.trimmed.fastq.gz"),
            sample=SAMPLES, read=READS
        ),
        # Bowtie2 alignment output files
        expand(path.join(ALIGN_DIR, "{sample}.sorted.bam"), sample=SAMPLES),
        expand(path.join(ALIGN_DIR, "{sample}.sorted.bam.bai"), sample=SAMPLES),
        # Peaks called by MACS2
        expand(
            path.join(PEAK_DIR, "{sample}_peaks.narrowPeak.gz"),
            sample=SAMPLES
        ),


